File: backend-py/api/business/product/product.py
from fastapi import APIRouter, HTTPException, Header
from pydantic import BaseModel, field_validator
from database import get_conn
from datetime import datetime
import pytz
import uuid
from jwt.jwt import decode_access_token
from typing import Optional

router = APIRouter(prefix="/api/business/product", tags=["Product"])

# Pydantic Model
class ProductCreateRequest(BaseModel):
    biz_id: str
    bp_barid: Optional[str] = None
    bp_sku: Optional[str] = None
    bp_pname: str
    bp_price: Optional[float] = None
    bp_cost: Optional[float] = None
    pc_id: str
    pu_id: str
    bp_nowstock: Optional[float] = None
    bp_stock_enable: Optional[int] = 0
    bp_descr: Optional[str] = None

    @field_validator("bp_stock_enable")
    @classmethod
    def validate_stock_enable(cls, v):
        if v is None:
            return 0
        if v not in [0, 1]:
            raise ValueError("bp_stock_enable must be 0 or 1")
        return v


@router.post("/create")
def create_product(
    data: ProductCreateRequest,
    authorization: str = Header(...)
):
    if not authorization.startswith("Bearer "):
        raise HTTPException(status_code=401, detail="Invalid token format")
    token = authorization.split(" ")[1]
    try:
        payload = decode_access_token(token)
    except Exception:
        raise HTTPException(status_code=401, detail="Token invalid")

    usr_id = payload.get("usr_id")
    if not usr_id:
        raise HTTPException(status_code=401, detail="Invalid user")

    if not data.biz_id or not data.bp_pname or not data.pc_id or not data.pu_id:
        raise HTTPException(status_code=400, detail="Missing required fields")

    try:
        if data.bp_price is not None:
            data.bp_price = float(data.bp_price)
        if data.bp_cost is not None:
            data.bp_cost = float(data.bp_cost)
        if data.bp_nowstock is not None:
            data.bp_nowstock = float(data.bp_nowstock)
        if data.bp_stock_enable is not None:
            data.bp_stock_enable = int(data.bp_stock_enable)
    except Exception:
        raise HTTPException(status_code=400, detail="Invalid data type for number fields")

    # --- Check for duplicate barcode or sku ---
    try:
        with get_conn() as conn:
            with conn.cursor() as cur:
                # Check barcode ซ้ำ
                if data.bp_barid:
                    cur.execute(
                        "SELECT bp_id FROM business_products WHERE biz_id = %s AND bp_barid = %s",
                        (data.biz_id, data.bp_barid)
                    )
                    if cur.fetchone():
                        raise HTTPException(status_code=400, detail="Barcode นี้ถูกใช้ไปแล้วในระบบ")
                # Check SKU ซ้ำ
                if data.bp_sku:
                    cur.execute(
                        "SELECT bp_id FROM business_products WHERE biz_id = %s AND bp_sku = %s",
                        (data.biz_id, data.bp_sku)
                    )
                    if cur.fetchone():
                        raise HTTPException(status_code=400, detail="SKU นี้ถูกใช้ไปแล้วในระบบ")
    except HTTPException as e:
        # re-raise ให้ FastAPI จัดการ
        raise e
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Database check failed: {e}")

    # --- Prepare to insert ---
    bp_id = str(uuid.uuid4())
    tz = pytz.timezone("Asia/Bangkok")
    now = datetime.now(tz)
    bp_added_ts = now.strftime("%Y-%m-%d %H:%M:%S")
    bp_last_update = bp_added_ts

    try:
        with get_conn() as conn:
            with conn.cursor() as cur:
                sql = """
                INSERT INTO business_products (
                    bp_id, biz_id, bp_barid, bp_sku, bp_pname, bp_price, bp_cost, pc_id,
                    pu_id, bp_nowstock, bp_stock_enable, bp_descr, bp_added_ts, bp_last_update
                ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
                """
                cur.execute(sql, (
                    bp_id, data.biz_id, data.bp_barid, data.bp_sku, data.bp_pname, data.bp_price,
                    data.bp_cost, data.pc_id, data.pu_id, data.bp_nowstock,
                    data.bp_stock_enable, data.bp_descr, bp_added_ts, bp_last_update
                ))
                conn.commit()
    except Exception as e:
        raise HTTPException(status_code=500, detail="Insert failed")

    return {
        "success": True,
        "bp_id": bp_id,
        "added_at": bp_added_ts
    }



@router.get("/category/getall/{bizid}")
def get_categories(bizid: str):
    if not bizid:
        return {"success": False, "data": [], "message": "Missing biz_id"}
    try:
        with get_conn() as conn:
            with conn.cursor() as cur:
                cur.execute("SELECT pc_id, pc_name FROM product_categories WHERE biz_id = %s", (bizid,))
                rows = cur.fetchall()
        return {"success": True, "data": rows}
    except Exception as e:
        return {"success": False, "data": [], "message": f"Database error: {e}"}


@router.get("/unit/getall/{bizid}")
def get_units(bizid: str):
    if not bizid:
        return {"success": False, "data": [], "message": "Missing biz_id"}
    try:
        with get_conn() as conn:
            with conn.cursor() as cur:
                cur.execute("SELECT pu_id, pu_name FROM product_units WHERE biz_id = %s", (bizid,))
                rows = cur.fetchall()
        return {"success": True, "data": rows}
    except Exception as e:
        return {"success": False, "data": [], "message": f"Database error: {e}"}

File: backend/api/business/product/product.js
const express = require('express');
const router = express.Router();
const { pool } = require('../../../database/database');
const { decodeAccessToken } = require('../../../function/jwt/jwt');
const { v4: uuidv4 } = require('uuid');

// --------- Create Product ----------
router.post('/create', async (req, res) => {
  const authorization = req.headers['authorization'];
  if (!authorization || !authorization.startsWith('Bearer ')) {
    return res.status(401).json({ detail: 'Invalid token format' });
  }

  let payload;
  try {
    payload = decodeAccessToken(authorization.split(' ')[1]);
  } catch (err) {
    return res.status(401).json({ detail: 'Token invalid' });
  }

  const usr_id = payload.usr_id;
  const data = req.body;

  // --- Validation (required) ---
  if (!usr_id) return res.status(401).json({ detail: 'Invalid user' });
  if (!data.biz_id || !data.bp_pname || !data.pc_id || !data.pu_id) {
    return res.status(400).json({ detail: 'Missing required fields' });
  }

  // --- Validate/convert numeric fields ---
  try {
    if (data.bp_price !== undefined && data.bp_price !== null) data.bp_price = parseFloat(data.bp_price);
    if (data.bp_cost !== undefined && data.bp_cost !== null) data.bp_cost = parseFloat(data.bp_cost);
    if (data.bp_nowstock !== undefined && data.bp_nowstock !== null) data.bp_nowstock = parseFloat(data.bp_nowstock);
    if (data.bp_stock_enable !== undefined && data.bp_stock_enable !== null) {
      data.bp_stock_enable = parseInt(data.bp_stock_enable, 10);
      if (![0, 1].includes(data.bp_stock_enable)) throw new Error();
    } else {
      data.bp_stock_enable = 0;
    }
  } catch {
    return res.status(400).json({ detail: 'Invalid data type for number fields' });
  }

  // --- Duplicate check: barcode/SKU ใน biz เดียวกันเท่านั้น ---
  try {
    if (data.bp_barid) {
      const [dupBar] = await pool.query(
        "SELECT bp_id FROM business_products WHERE biz_id = ? AND bp_barid = ?",
        [data.biz_id, data.bp_barid]
      );
      if (dupBar.length > 0) {
        return res.status(400).json({ detail: "Barcode นี้ถูกใช้ไปแล้วในระบบ" });
      }
    }
    if (data.bp_sku) {
      const [dupSku] = await pool.query(
        "SELECT bp_id FROM business_products WHERE biz_id = ? AND bp_sku = ?",
        [data.biz_id, data.bp_sku]
      );
      if (dupSku.length > 0) {
        return res.status(400).json({ detail: "SKU นี้ถูกใช้ไปแล้วในระบบ" });
      }
    }
  } catch (e) {
    return res.status(500).json({ detail: `Database check failed: ${e.message}` });
  }

  // --- Prepare for insert ---
  const bp_id = uuidv4();
  const now = new Date();
  const bp_added_ts = now.toLocaleString("sv-SE", { timeZone: "Asia/Bangkok" }).replace('T', ' ').slice(0, 19);
  const bp_last_update = bp_added_ts;

  try {
    await pool.query(
      `INSERT INTO business_products (
        bp_id, biz_id, bp_barid, bp_sku, bp_pname, bp_price, bp_cost, pc_id,
        pu_id, bp_nowstock, bp_stock_enable, bp_descr, bp_added_ts, bp_last_update
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      [
        bp_id, data.biz_id, data.bp_barid || null, data.bp_sku || null, data.bp_pname,
        data.bp_price || null, data.bp_cost || null, data.pc_id, data.pu_id,
        data.bp_nowstock || null, data.bp_stock_enable, data.bp_descr || null,
        bp_added_ts, bp_last_update
      ]
    );
    return res.json({
      success: true,
      bp_id,
      added_at: bp_added_ts
    });
  } catch (e) {
    return res.status(500).json({ detail: "Insert failed" });
  }
});

// --------- Get Product Categories ----------
router.get('/category/getall/:bizid', async (req, res) => {
  const bizid = req.params.bizid;
  if (!bizid) {
    return res.json({ success: false, data: [], message: "Missing biz_id" });
  }
  try {
    const [rows] = await pool.query(
      "SELECT pc_id, pc_name FROM product_categories WHERE biz_id = ?",
      [bizid]
    );
    return res.json({ success: true, data: rows });
  } catch (e) {
    return res.json({ success: false, data: [], message: `Database error: ${e.message}` });
  }
});

// --------- Get Product Units ----------
router.get('/unit/getall/:bizid', async (req, res) => {
  const bizid = req.params.bizid;
  if (!bizid) {
    return res.json({ success: false, data: [], message: "Missing biz_id" });
  }
  try {
    const [rows] = await pool.query(
      "SELECT pu_id, pu_name FROM product_units WHERE biz_id = ?",
      [bizid]
    );
    return res.json({ success: true, data: rows });
  } catch (e) {
    return res.json({ success: false, data: [], message: `Database error: ${e.message}` });
  }
});

module.exports = router;
